library/xml/src/modules/XML.rb

```diff
@@ -97,96 +119,236 @@ def xmlCreateDoc(doc, doc_settings)
+
+      result[name] = case type
+      # disksize is our own design mistake. deprecated.
+      when "string", "disksize" then text
+      when "symbol"
+        raise XMLDeserializationError.for_node(node, "empty symbols are invalid") if text.empty?
+
+        text.to_sym
+      when "integer"
+        raise XMLDeserializationError.for_node(node, "cannot be parsed as an integer") if text !~ /-?\d+/
+
+        text.to_i
+      when "boolean"
+        v = BOOLEAN_MAP[text]
+        raise XMLDeserializationError.for_node(node, "booleans can only be 'true' or 'false'") if v.nil?
+
+        v
+      when "list"
+        children.map do |kid|
+          # always pass new hash to prevent overwrite for list with same elements
+          r = {}
+          parse_node(kid, r)
+          r.values.first
+        end
+      when "map"
+        v = {}
+        children.each { |kid| parse_node(kid, v) }
+        v
+      else
+        raise XMLDeserializationError.for_node(node, "invalid type #{type.inspect}")
+      end
+
+      result
+    end
+
+    # @param [Nokogiri::XML::Document] doc
+    # @param [Hash] metadata for current doc
+    # @param [Nokogiri::XML::Node] parent
+    # @param [Hash] contents content to write
+    # @param [Array] parent_array (for better error reporting)
+    # @return [void]
+    def add_element(doc, metadata, parent, contents, parent_array = nil)
+      # backward compatibility. Keys are sorted and needs old ycp sort to be able to compare also classes
+      Builtins.sort(contents.keys).each do |key|
+        raise XMLSerializationError.new("Cannot represent non-string key '#{key.inspect}', part of #{contents.inspect}", contents) unless key.is_a?(::String)
+
+        value = contents[key]
+        type_attr = "t"
+        element = Nokogiri::XML::Node.new(key, doc)
+        case value
+        when ::String
+          element.content = value
+        when ::Integer
+          element[type_attr] = "integer"
+          element.content = value.to_s
+        when ::Symbol
+          element[type_attr] = "symbol"
+          element.content = value.to_s
+        when true, false
+          element[type_attr] = "boolean"
+          element.content = value.to_s
+        when ::Array
+          element[type_attr] = "list"
+          special_names = metadata["listEntries"] || {}
+          element_name = special_names[key] || "listentry"
+          value.each do |list_value|
+            add_element(doc, metadata, element, { element_name => list_value }, value)
+          end
+        when ::Hash
+          element[type_attr] = "map"
+          add_element(doc, metadata, element, value)
+        else # including nil
+          # without parent_array, a ["foo", nil, "bar"] would always
+          # unhelpfully report {"listentry"=> nil} as the parent object
+          o = parent_array || contents
+          raise XMLSerializationError.new("Cannot represent #{value.inspect}, part of #{o.inspect}", o)
```

Comment:
I worry that even that part can be useless if it is something like "[nil]". What we probably should do is always log whole structure we serialize and raise exception. As I think it is always our fault to provide wrong ycp value to serializer, so it always will need logs and our fix.

Reply:
I think it is reasonably balanced now, when we haven't checked all the usage sites. If we do that, we can improve the library further.